### PR TITLE
Parallel Tuning Strategies - MPI Communication for all existing tuning strategies

### PR DESCRIPTION
**Pull request by wothie**
# Description

Provides a new TuningStrategy (MPIParallelizedStrategy) which wraps around another TuningStrategy and handles its communication to other ranks. More specifically it splits the provided options among all ranks, lets them tune locally according to their strategy and then selects the optimal configuration across all ranks.
For this purpose it includes a new compiler option, new Options (MPIStrategyOption) and a wrapper for basic mpi calls to reduce the number of ifdefs. It also provides the AutoPasConfigurationCommunicator.h to handle distributing and comparing configurations across ranks, which is used in MPIParallelizedStrategy.
Due to how communication is done, it is pretty much limited to a homogeneous domain.

## Todo
- [x] Adapt ContributionGuidelines if needed
- [X] Try implementing Active Harmony without MPI. Result: Doesn't work as client count needs to be set during session start.

## Related Pull Requests

- Basically the completed version of #446 

## Resolved Issues

- [X] #410 mostly (except for the heterogeneous case)

# How Has This Been Tested?

- [X] locally with newly written tests and mostly sph-main-mpi
- [x] with all examples
- [x] on a cluster
- [x] very thoroughly

### CODE REVIEW

**Comment by SteffenSeckler** [comment]
@wothie you have somehow broken the tests `Generated/TuningStrategyTest.testSearchSpaceOneOption/*`

File: tests/MPIParallelAutoPasTests/tests/selectors/tuningStrategy/MPIParallelizedStrategyTest.cpp [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  auto strategy = std::make_unique<FullSearch>(oneContainer, twoCellSizes, oneTraversal, oneLoadEstimator,
+                                               oneDataLayout, oneNewton3);
+
+  auto mpiParallelizedStrategy = MPIParallelizedStrategy(std::move(strategy), MPI_COMM_WORLD, oneContainer,
+                                                         oneTraversal, oneLoadEstimator, oneDataLayout, oneNewton3);
+  finiteCellSizeFactorsSetup(mpiParallelizedStrategy);
+
+  EXPECT_EQ(mpiParallelizedStrategy.getCurrentConfiguration(),
+            Configuration(ContainerOption::linkedCells, 1.0, TraversalOption::lc_sliced, LoadEstimatorOption::none,
+                          DataLayoutOption::aos, Newton3Option::enabled));
```
Comment (SteffenSeckler) [inline tests/MPIParallelAutoPasTests/tests/selectors/tuningStrategy/MPIParallelizedStrategyTest.cpp:64]: why? describe as comment

File: tests/MPIParallelAutoPasTests/tests/selectors/tuningStrategy/MPIParallelizedStrategyTest.cpp [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  const auto oneNewton3 = std::set<Newton3Option>{Newton3Option::enabled};
+
+  auto strategy = std::make_unique<RandomSearch>(oneContainer, twoCellSizes, oneTraversal, oneLoadEstimator,
+                                                 oneDataLayout, oneNewton3, 2);
+
+  auto mpiParallelizedStrategy = MPIParallelizedStrategy(std::move(strategy), MPI_COMM_WORLD, oneContainer,
+                                                         oneTraversal, oneLoadEstimator, oneDataLayout, oneNewton3);
+  finiteCellSizeFactorsSetup(mpiParallelizedStrategy);
+
+  EXPECT_LE(mpiParallelizedStrategy.getCurrentConfiguration().cellSizeFactor, 1.1);
```
Comment (SteffenSeckler) [inline tests/MPIParallelAutoPasTests/tests/selectors/tuningStrategy/MPIParallelizedStrategyTest.cpp:85]: explain, why.

File: tests/MPIParallelAutoPasTests/tests/selectors/tuningStrategy/MPIParallelizedStrategyTest.cpp [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  const auto oneNewton3 = std::set<Newton3Option>{Newton3Option::enabled};
+
+  auto strategy = std::make_unique<RandomSearch>(oneContainer, twoCellSizes, oneTraversal, oneLoadEstimator,
+                                                 oneDataLayout, oneNewton3, 2);
+
+  auto mpiParallelizedStrategy = MPIParallelizedStrategy(std::move(strategy), MPI_COMM_WORLD, oneContainer,
+                                                         oneTraversal, oneLoadEstimator, oneDataLayout, oneNewton3);
+  auto smallestLocalCellSizeFactor = infiniteCellSizeFactorSetup(mpiParallelizedStrategy);
+
+  EXPECT_LE(mpiParallelizedStrategy.getCurrentConfiguration().cellSizeFactor, smallestLocalCellSizeFactor);
```
Comment (SteffenSeckler) [inline tests/MPIParallelAutoPasTests/tests/selectors/tuningStrategy/MPIParallelizedStrategyTest.cpp:106]: explain why

File: tests/MPIParallelAutoPasTests/tests/selectors/tuningStrategy/MPIParallelizedStrategyTest.cpp [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  const auto oneNewton3 = std::set<Newton3Option>{Newton3Option::enabled};
+
+  auto strategy = std::make_unique<ActiveHarmony>(oneContainer, twoCellSizes, oneTraversal, oneLoadEstimator,
+                                                  oneDataLayout, oneNewton3);
+
+  auto mpiParallelizedStrategy = MPIParallelizedStrategy(std::move(strategy), MPI_COMM_WORLD, oneContainer,
+                                                         oneTraversal, oneLoadEstimator, oneDataLayout, oneNewton3);
+  finiteCellSizeFactorsSetup(mpiParallelizedStrategy);
+
+  EXPECT_EQ(mpiParallelizedStrategy.getCurrentConfiguration(),
```
Comment (SteffenSeckler) [inline tests/MPIParallelAutoPasTests/tests/selectors/tuningStrategy/MPIParallelizedStrategyTest.cpp:127]: explain why

File: tests/MPIParallelAutoPasTests/tests/selectors/tuningStrategy/MPIParallelizedStrategyTest.cpp [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+  auto strategy = std::make_unique<BayesianSearch>(oneContainer, twoCellSizes, oneTraversal, oneLoadEstimator,
+                                                   oneDataLayout, oneNewton3);
+
+  auto mpiParallelizedStrategy = MPIParallelizedStrategy(std::move(strategy), MPI_COMM_WORLD, oneContainer,
+                                                         oneTraversal, oneLoadEstimator, oneDataLayout, oneNewton3);
+  auto smallestCellSizeFactor = infiniteCellSizeFactorSetup(mpiParallelizedStrategy);
+
+  // BayesianSearch seems to not be absolutely optimal in this case.
+  double error = 0.01;
+  EXPECT_LE(mpiParallelizedStrategy.getCurrentConfiguration().cellSizeFactor, smallestCellSizeFactor + error);
```
Comment (SteffenSeckler) [inline tests/MPIParallelAutoPasTests/tests/selectors/tuningStrategy/MPIParallelizedStrategyTest.cpp:152]: why?

File: tests/MPIParallelAutoPasTests/tests/utils/AutoPasConfigurationCommunicatorTest.cpp [hunk 6]
```diff
@@ -0,0 +1,219 @@
+/**
```
Comment (SteffenSeckler) [inline tests/MPIParallelAutoPasTests/tests/utils/AutoPasConfigurationCommunicatorTest.cpp:1]: shortly explain each test:
what are you testing and why do you expect, what you checking.

**Review by SteffenSeckler — CHANGES_REQUESTED** [review]
see comments.

File: src/autopas/selectors/tuningStrategy/ActiveHarmony.h [hunk 7]
```diff
@@ -6,12 +6,21 @@
 
 #pragma once
 
+#include <unistd.h>
```
[the remark this answers is not in the document]
Comment (SteffenSeckler) [reply src/autopas/selectors/tuningStrategy/ActiveHarmony.h:9]: you can find `getenv` in `cstdlib` and `snprintf` in `cstdio`

File: src/autopas/selectors/tuningStrategy/BayesianClusterSearch.h [hunk 8]
```diff
@@ -166,9 +167,14 @@ class BayesianClusterSearch : public TuningStrategyInterface {
...
     ++_currentNumEvidence;
     _currentAcquisitions.clear();
+
+    _traversalTimes[_currentConfig] = time;
   }
 
+  inline long getEvidence(Configuration configuration) const override { return _traversalTimes.at(configuration); }
+
   inline void reset(size_t iteration) override {
+    _gaussianCluster.clear();
```
[the remark this answers is not in the document]
Comment (Njan0) [reply src/autopas/selectors/tuningStrategy/BayesianClusterSearch.h:177]: This is intended. GaussianCluster needs a bunch of evidence to work properly. We keep the evidence over several tuning phases to avoid constant resampling.
```suggestion
```

[the remark this answers is not in the document]
Comment (SteffenSeckler) [reply src/autopas/selectors/tuningStrategy/BayesianClusterSearch.h:177]: when do you actually clear this?

[the remark this answers is not in the document]
Comment (Njan0) [reply src/autopas/selectors/tuningStrategy/BayesianClusterSearch.h:177]: Currently we only decay their influence over time. Theoretically we could delete them if their influence becomes negligible. But we have not yet encountered any problems in our tests yet. To cause performance problems you probably have to set maxEvidence very high.

However, if this is a problem, the best option would probably be to restart the whole strategy after a few tuning phases.

```suggestion
    if (++phasesSinceLastReset >= phasesPerReset) {
        _gaussianCluster.clear();
        _firstTuningPhase = true;
        phasesSinceLastReset = 0;
    }
```

[the remark this answers is not in the document]
Comment (Njan0) [reply src/autopas/selectors/tuningStrategy/BayesianClusterSearch.h:177]: On that note you probably need to add the following after resetting fullSearch.

`_currentConfig = _fullSearch.getCurrentConfiguration();`

https://github.com/AutoPas/AutoPas/blob/b2f476522ed55acd098edffd951066d581e918c1/src/autopas/selectors/tuningStrategy/BayesianClusterSearch.h#L188-L190

File: src/autopas/selectors/tuningStrategy/ActiveHarmony.h [hunk 9]
```diff
@@ -29,30 +38,33 @@ class ActiveHarmony : public TuningStrategyInterface {
...
-    AutoPasLog(debug, "Reducing traversal and container options to possible combinations");
-    for (auto &traversalOption : _allowedTraversalOptions) {
-      auto compatibleContainers = compatibleTraversals::allCompatibleContainers(traversalOption);
-      if (compatibleContainers.empty() or
-          allowedContainerOptions.find(*compatibleContainers.begin()) == allowedContainerOptions.end()) {
-        _allowedTraversalOptions.erase(traversalOption);
-      } else {
-        _allowedContainerOptions.emplace(*compatibleContainers.begin());
-      }
-    }
```
[the remark this answers is not in the document]
Comment (wothie) [reply src/autopas/selectors/tuningStrategy/ActiveHarmony.h:55]: Got replaced by a call to utils::AutoPasConfigurationCommunicator::distributeConfigurations(...), which essentially does the same thing.

File: src/autopas/selectors/tuningStrategy/RandomSearch.h [hunk 10]
```diff
@@ -46,23 +46,49 @@ class RandomSearch : public TuningStrategyInterface {
...
   inline void removeN3Option(Newton3Option badNewton3Option) override;
 
-  inline void addEvidence(long time, size_t iteration) override { _traversalTimes[_currentConfig] = time; }
+  inline void addEvidence(long time, size_t iteration) override {
+    auto testingConfig = Configuration(_currentConfig);
+    testingConfig.cellSizeFactor = 1;
+    if (_traversalTimes.find(testingConfig) == _traversalTimes.end()) {
+      ++_numTestedConfigsNoCSF;
+    }
+    _traversalTimes[_currentConfig] = time;
```
[the remark this answers is not in the document]
Comment (wothie) [reply src/autopas/selectors/tuningStrategy/RandomSearch.h:83]: Came from a conversation I had with @FG-TUM 
Essentially I couldn't find a reliable way to figure out how many valid configs there could still exist for infinite CSFs, so I ignore them alltogether.
This is not perfect and we might want to have another talk about a cleaner solution at some point.

[the remark this answers is not in the document]
Comment (SteffenSeckler) [reply src/autopas/selectors/tuningStrategy/RandomSearch.h:83]: can you document this? -> write a comment.

File: src/autopas/selectors/tuningStrategy/MPIParallelizedStrategy.h [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+
+  inline void addEvidence(long time, size_t iteration) override { _tuningStrategy->addEvidence(time, iteration); }
+
+  [[nodiscard]] inline long getEvidence(Configuration configuration) const override {
+    return _tuningStrategy->getEvidence(configuration);
+  }
+
+  [[nodiscard]] const Configuration &getCurrentConfiguration() const override {
+    // cellSizeFactor == -1 iff a config is invalid
+    if (_optimalConfiguration.cellSizeFactor == -1) {
```
[the remark this answers is not in the document]
Comment (SteffenSeckler) [reply src/autopas/selectors/tuningStrategy/MPIParallelizedStrategy.h:60]: Maybe you could add the method `isValid()` to the Configuration class. Might help to make this clearer.

File: src/autopas/selectors/tuningStrategy/BayesianClusterSearch.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (wothie) [reply src/autopas/selectors/tuningStrategy/BayesianClusterSearch.h]: @Njan0 do I understand it correctly that the last two suggestions are not really necessary?
I just assumed that the gaussianCluster should also be reset, but if you say there are no problems by not calling clear() than I don't see any either.
Is your last comment also related to this or is it a general suggestion?

File: src/autopas/selectors/tuningStrategy/MPIParallelizedStrategy.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (wothie) [reply src/autopas/selectors/tuningStrategy/MPIParallelizedStrategy.h:60]: Do you think ```hasValidValues()``` would get the point across properly?

[the remark this answers is not in the document]
Comment (SteffenSeckler) [reply src/autopas/selectors/tuningStrategy/MPIParallelizedStrategy.h:60]: seems fine

File: src/autopas/selectors/tuningStrategy/BayesianClusterSearch.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (Njan0) [reply src/autopas/selectors/tuningStrategy/BayesianClusterSearch.h]: My last comment would generally be the correct implementation. But this is only a problem if someone calls reset() although tune() still returns true. So something that currently should never happen. But maybe that will change one day. ;)

File: src/autopas/selectors/tuningStrategy/ActiveHarmony.h [hunk 12]
```diff
@@ -29,30 +38,33 @@ class ActiveHarmony : public TuningStrategyInterface {
    * @param allowedLoadEstimatorOptions
    * @param allowedDataLayoutOptions
    * @param allowedNewton3Options
+   * @param mpiStrategyOption
+   * @param comm Default value shouldn't be used. Only provided to not have to set it as first parameter.
    */
   ActiveHarmony(const std::set<ContainerOption> &allowedContainerOptions = ContainerOption::getAllOptions(),
```
[the remark this answers is not in the document]
Comment (wothie) [reply src/autopas/selectors/tuningStrategy/ActiveHarmony.h:44]: The tests also call the constructor. Since I generally use ```mpi.h``` in the tests and ```MPIWrapper.h``` inside of AutoPas, I need the default values to be able to call the constructor at all.

[the remark this answers is not in the document]
Comment (SteffenSeckler) [reply src/autopas/selectors/tuningStrategy/ActiveHarmony.h:47]: Just tested this:
removing ALL defaults for this constructor and adding MPI_COMM_WORLD (or AUTOPAS_MPI_COMM_WORLD) as last argument for the constructor (make_unique) in `MPIParallelizedStrategyTest.cpp` works.

**Comment by SteffenSeckler** [comment]
fyi: if you merge the same remote branch into your own branch, it is often recommended to use rebase instead of merge.
This will apply the remote changes first and replay your local changes/commits afterward.

**Comment by wothie** [comment]
@SteffenSeckler the review suggestions have all been addressed so you might look over the non-trivial things again, if you want to.
(is there a "request review of changes since last review"-button somewhere?)

**Comment by SteffenSeckler** [comment]
@wothie there's a re-request review at the top ;)

**Review by SteffenSeckler — APPROVED** [review]
only a few minor things.
Please adapt them first. Then we can merge the PR.

File: src/autopas/selectors/tuningStrategy/ActiveHarmony.h [hunk 13]
```diff
@@ -361,7 +364,7 @@ void ActiveHarmony::resetHarmony() {
       utils::ExceptionHandler::exception("ActiveHarmony::reset: Error connecting to Harmony session: {}", ah_error());
     }
 
-    if (!_nonLocalServer || rank == 0) {
+    if (not _nonLocalServer || rank == 0) {
```
[the remark this answers is not in the document]
Comment (wothie) [reply src/autopas/selectors/tuningStrategy/ActiveHarmony.h:367]: Should I also change other uses of ```||``` and ```&&``` to ```or``` and ```and```?

[the remark this answers is not in the document]
Comment (SteffenSeckler) [reply src/autopas/selectors/tuningStrategy/ActiveHarmony.h:367]: if you find any, please do :)
